vcluster panic when connecting using --kube-config-context-name

### What happened?

Trying to connect to a vcluster.

the following code works
```
vcluster \
    --context   "${KUBE_CLU_NAME}" \
    --namespace "${VCLU_NS}" \
    connect "${VCLU_NAME}" \
    --server https://${VCLU_NAME}.${VCLU_DNS_SUFFIX} \
```

but if we add --kube-config-context-name it fails
```
vcluster \
    --context   "${KUBE_CLU_NAME}" \
    --namespace "${VCLU_NS}" \
    connect "${VCLU_NAME}" \
    --server https://${VCLU_NAME}.${VCLU_DNS_SUFFIX} \
    --kube-config-context-name "${VCLU_NAME}"
```

```
vcluster --context eks-cluster-01 --namespace v-clu-1-name connect v-clu-1-name --server https://v-clu-1-name.somedomain.someext --kube-config-context-name v-clu-1-name
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x103c3101c]

goroutine 1 [running]:
k8s.io/client-go/tools/clientcmd.ModifyConfig({0x104825080, 0x1400011a9b0}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x1400015ff80}, 0x1400015ffb0, 0x140005420c0, ...}, ...)
        k8s.io/client-go@v0.26.1/tools/clientcmd/config.go:208 +0x38c
github.com/loft-sh/vcluster/cmd/vclusterctl/cmd.updateKubeConfig({0x16d2bf719, 0x14}, 0x0, 0x0, 0x1)
        github.com/loft-sh/vcluster/cmd/vclusterctl/cmd/util.go:143 +0x328
github.com/loft-sh/vcluster/cmd/vclusterctl/cmd.(*ConnectCmd).Connect(0x140005afe00, {0x16d2bf6a6, 0x14}, {0x1400014ba80, 0x0, 0x8})
        github.com/loft-sh/vcluster/cmd/vclusterctl/cmd/connect.go:181 +0x784
github.com/loft-sh/vcluster/cmd/vclusterctl/cmd.(*ConnectCmd).Run(...)
        github.com/loft-sh/vcluster/cmd/vclusterctl/cmd/connect.go:128
github.com/loft-sh/vcluster/cmd/vclusterctl/cmd.NewConnectCmd.func1(0x14000622c00?, {0x1400014ba70, 0x1, 0x9})
        github.com/loft-sh/vcluster/cmd/vclusterctl/cmd/connect.go:101 +0x88
github.com/spf13/cobra.(*Command).execute(0x14000622c00, {0x1400014b9e0, 0x9, 0x9})
        github.com/spf13/cobra@v1.6.0/command.go:916 +0x5c8
github.com/spf13/cobra.(*Command).ExecuteC(0x14000622900)
        github.com/spf13/cobra@v1.6.0/command.go:1040 +0x354
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.6.0/command.go:968
github.com/loft-sh/vcluster/cmd/vclusterctl/cmd.Execute()
        github.com/loft-sh/vcluster/cmd/vclusterctl/cmd/root.go:47 +0xbc
main.main()
        ./main.go:17 +0x30
```


### What did you expect to happen?

the command to work, and add an entry to ~/.kube/config with the designated context

### How can we reproduce it (as minimally and precisely as possible)?

(see description above)

### Anything else we need to know?

* CLI: vcluster 0.15.0 on macOS ventura (arm m1)
* Server: EKS 1.26
* VCluster: with vcluster-eks 0.15.0 deployed via helm chart (from flux2)

### Host cluster Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-14T13:14:41Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"26+", GitVersion:"v1.26.2-eks-a59e1f0", GitCommit:"8b68f4b95d7121d039ceebd30870e48acc7772e4", GitTreeState:"clean", BuildDate:"2023-03-09T19:59:45Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
```

</details>


### Host cluster Kubernetes distribution

<details>

```
EKS 1.26
```

</details>


### vlcuster version

<details>

```console
$ vcluster --version
vcluster version 0.15.0
```

</details>


### Vcluster Kubernetes distribution(k3s(default)), k8s, k0s)

<details>

```
eks
```

</details>


### OS and Arch

<details>

```
OS: linux
Arch: amd64
```
</details>

--- a/cmd/vclusterctl/cmd/connect.go
+++ b/cmd/vclusterctl/cmd/connect.go
@@ -500,6 +500,9 @@ func (cmd *ConnectCmd) setServerIfExposed(vClusterName string, vClusterConfig *a
 	return nil
 }
 
+// exchangeContextName switches the context name specified in the remote kubeconfig with
+// the context name specified by the user. It cannot correctly handle kubeconfigs with multiple entries
+// for clusters, authInfos, contexts, but ideally this is pointed at a secret created by us.
 func (cmd *ConnectCmd) exchangeContextName(kubeConfig *api.Config, vclusterName string) error {
 	if cmd.KubeConfigContextName == "" {
 		if vclusterName != "" {
@@ -509,27 +512,36 @@ func (cmd *ConnectCmd) exchangeContextName(kubeConfig *api.Config, vclusterName 
 		}
 	}
 
-	// update cluster
+	// pick the last specified cluster (there should ideally be exactly one)
 	for k := range kubeConfig.Clusters {
 		kubeConfig.Clusters[cmd.KubeConfigContextName] = kubeConfig.Clusters[k]
-		delete(kubeConfig.Clusters, k)
+		// delete the rest
+		if k != cmd.KubeConfigContextName {
+			delete(kubeConfig.Clusters, k)
+		}
 		break
 	}
 
-	// update context
+	// pick the last specified context (there should ideally be exactly one)
 	for k := range kubeConfig.Contexts {
 		ctx := kubeConfig.Contexts[k]
 		ctx.Cluster = cmd.KubeConfigContextName
 		ctx.AuthInfo = cmd.KubeConfigContextName
 		kubeConfig.Contexts[cmd.KubeConfigContextName] = ctx
-		delete(kubeConfig.Contexts, k)
+		// delete the rest
+		if k != cmd.KubeConfigContextName {
+			delete(kubeConfig.Contexts, k)
+		}
 		break
 	}
 
-	// update authInfo
+	// pick the last specified authinfo (there should ideally be exactly one)
 	for k := range kubeConfig.AuthInfos {
 		kubeConfig.AuthInfos[cmd.KubeConfigContextName] = kubeConfig.AuthInfos[k]
-		delete(kubeConfig.AuthInfos, k)
+		// delete the rest
+		if k != cmd.KubeConfigContextName {
+			delete(kubeConfig.AuthInfos, k)
+		}
 		break
 	}
 

--- a/cmd/vclusterctl/cmd/connect_test.go
+++ b/cmd/vclusterctl/cmd/connect_test.go
@@ -1,0 +1,98 @@
+package cmd
+
+import (
+	"testing"
+
+	"github.com/loft-sh/vcluster/cmd/vclusterctl/flags"
+	"gotest.tools/v3/assert"
+	"k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	defaultContextName = "my-vcluster"
+)
+
+func TestExchangeContextName(t *testing.T) {
+	vclusterName := "vcluster-name"
+	namespace := "default"
+	defaultContext := &api.Config{
+		Clusters: map[string]*api.Cluster{
+			defaultContextName: {Server: "foo"},
+		},
+		AuthInfos: map[string]*api.AuthInfo{
+			defaultContextName: {
+				Token: "foo",
+			},
+		},
+		Contexts: map[string]*api.Context{
+			defaultContextName: {
+				Cluster:  defaultContextName,
+				AuthInfo: defaultContextName,
+			},
+		},
+		CurrentContext: defaultContextName,
+	}
+	testTable := []struct {
+		desc           string
+		newContextName string
+		config         *api.Config
+		expectedConfig *api.Config
+	}{
+		{
+			desc:           "KubeConfigContextName specified",
+			newContextName: "new-context",
+			config:         defaultContext.DeepCopy(),
+			expectedConfig: &api.Config{
+				Clusters: map[string]*api.Cluster{
+					"new-context": {Server: "foo"},
+				},
+				AuthInfos: map[string]*api.AuthInfo{
+					"new-context": {
+						Token: "foo",
+					},
+				},
+				Contexts: map[string]*api.Context{
+					"new-context": {
+						Cluster:  "new-context",
+						AuthInfo: "new-context",
+					},
+				},
+				CurrentContext: "new-context",
+			},
+		},
+		{
+			desc:           "KubeConfigContextName same as default",
+			newContextName: defaultContextName,
+			config:         defaultContext.DeepCopy(),
+			expectedConfig: &api.Config{
+				Clusters: map[string]*api.Cluster{
+					defaultContextName: &api.Cluster{Server: "foo"},
+				},
+				AuthInfos: map[string]*api.AuthInfo{
+					defaultContextName: &api.AuthInfo{
+						Token: "foo",
+					},
+				},
+				Contexts: map[string]*api.Context{
+					defaultContextName: &api.Context{
+						Cluster:  defaultContextName,
+						AuthInfo: defaultContextName,
+					},
+				},
+				CurrentContext: defaultContextName,
+			},
+		},
+	}
+	for _, testCase := range testTable {
+		cmd := &ConnectCmd{
+			GlobalFlags: &flags.GlobalFlags{
+				Namespace: namespace,
+			},
+			KubeConfigContextName: testCase.newContextName,
+		}
+		assert.NilError(t, cmd.exchangeContextName(testCase.config, vclusterName))
+		newConfig := testCase.config.DeepCopy()
+
+		assert.DeepEqual(t, newConfig, testCase.expectedConfig)
+	}
+}

